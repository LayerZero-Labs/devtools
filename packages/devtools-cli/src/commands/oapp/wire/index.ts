import { Command } from 'commander'
import { printLogo } from '@layerzerolabs/io-devtools/swag'
import { createLogger, setDefaultLogLevel } from '@layerzerolabs/io-devtools'
import {
    type WithLogLevelOption,
    type WithSetupOption,
    WithTsConfigOption,
    createLogLevelOption,
    createSetupFileOption,
    createTsConfigFileOption,
} from '@/commands/options'
import { setupTypescript } from '@/setup'

interface Args extends WithLogLevelOption, WithSetupOption, WithTsConfigOption {}

export const wire = new Command('wire')
    .addOption(createSetupFileOption())
    .addOption(createTsConfigFileOption())
    .addOption(createLogLevelOption())
    .action(async ({ setup, logLevel, tsConfig: tsConfigPath }: Args) => {
        printLogo()

        // We'll set the global logging level to get as much info as needed
        setDefaultLogLevel(logLevel)

        // We'll setup TypeScript support so that we can dynamically load TypeScript config files
        setupTypescript(tsConfigPath)

        const logger = createLogger(logLevel)

        logger.debug(`Loading setup from ${setup}`)

        logger.warn(
            `This command is just a placeholder. Please use @layerzerolabs/toolbox-hardhat package for the time being.`
        )
    })
