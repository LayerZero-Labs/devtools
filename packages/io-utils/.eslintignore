.turbo
dist
node_modules