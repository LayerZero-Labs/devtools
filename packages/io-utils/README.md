<p align="center">
  <a href="https://layerzero.network">
    <img alt="LayerZero" style="max-width: 500px" src="https://d3a2dpnnrypp5h.cloudfront.net/bridge-app/lz.png"/>
  </a>
</p>

<h1 align="center">@layerzerolabs/io-utils</h1>

<!-- The badges section -->
<p align="center">
  <!-- Shields.io NPM published package version -->
  <a href="https://www.npmjs.com/package/@layerzerolabs/io-utils"><img alt="NPM Version" src="https://img.shields.io/npm/v/@layerzerolabs/io-utils"/></a>
  <!-- Shields.io NPM downloads -->
  <a href="https://www.npmjs.com/package/@layerzerolabs/io-utils"><img alt="Downloads" src="https://img.shields.io/npm/dm/@layerzerolabs/io-utils"/></a>
  <!-- Shields.io license badge -->
  <a href="https://www.npmjs.com/package/@layerzerolabs/io-utils"><img alt="NPM License" src="https://img.shields.io/npm/l/@layerzerolabs/io-utils"/></a>
</p>

## Installation

```bash
pnpm install @layerzerolabs/io-utils
```

```bash
yarn install @layerzerolabs/io-utils
```

```bash
npm install @layerzerolabs/io-utils
```
