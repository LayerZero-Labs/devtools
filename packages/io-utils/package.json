{
  "name": "@layerzerolabs/io-utils",
  "version": "0.0.1",
  "private": true,
  "description": "Utilities for working with I/O in LayerZero utils",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/LayerZero-Labs/lz-utils.git",
    "directory": "packages/io-utils"
  },
  "license": "MIT",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "require": "./dist/index.js",
      "import": "./dist/index.mjs"
    },
    "./*": {
      "types": "./dist/*.d.ts",
      "require": "./dist/*.js",
      "import": "./dist/*.mjs"
    }
  },
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "prebuild": "tsc -noEmit",
    "build": "npx tsup",
    "clean": "rm -rf dist",
    "lint": "npx eslint '**/*.{js,ts,json}'",
    "test": "jest --passWithNoTests"
  },
  "dependencies": {
    "cli-table3": "^0.6.3",
    "prompts": "^2.4.2",
    "winston": "^3.11.0"
  },
  "devDependencies": {
    "@types/jest": "^29.5.10",
    "fast-check": "^3.14.0",
    "jest": "^29.7.0",
    "ts-jest": "^29.1.1",
    "ts-node": "^10.9.1",
    "tslib": "~2.6.2",
    "tsup": "~8.0.1",
    "typescript": "^5.2.2",
    "zod": "^3.22.4"
  },
  "peerDependencies": {
    "zod": "^3.22.4"
  }
}
