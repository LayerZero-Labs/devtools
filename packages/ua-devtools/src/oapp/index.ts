export * from './check'
export * from './config'
export * from './schema'
export * from './types'
