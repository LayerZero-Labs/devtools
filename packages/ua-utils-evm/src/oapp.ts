import { Contract } from "@ethersproject/contracts"
import { GetPropertyValue, createProperty } from "@layerzerolabs/ua-utils"
import { EndpointId } from "@layerzerolabs/lz-definitions"

type SetPeerConfigurableContext = [oapp: Contract, endpointId: EndpointId]

type SetPeerConfigurableValue = string

export const createSetPeerProperty = (desired: GetPropertyValue<SetPeerConfigurableContext, SetPeerConfigurableValue>) =>
    createProperty<SetPeerConfigurableContext, SetPeerConfigurableValue>({
        desired,
        get: (oapp, endpointId) => oapp.peers(endpointId),
        set: (oapp, endpointId, peer) => oapp.setPeer(endpointId, peer),
    })
