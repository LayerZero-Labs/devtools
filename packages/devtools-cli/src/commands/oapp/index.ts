import { Command } from 'commander'
import { wire } from './wire'

export const oapp = new Command('oapp').description('OApp configuration commands').addCommand(wire)
