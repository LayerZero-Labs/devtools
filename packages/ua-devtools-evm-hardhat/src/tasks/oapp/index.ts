import './wire'
import './peers.get'
import './config.get.default'
import './config.get.executor'
import './config.get'
import './enforced.opts.get'
import './config.init'
import './subtask.config.load'
