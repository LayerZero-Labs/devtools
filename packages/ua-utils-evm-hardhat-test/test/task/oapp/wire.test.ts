import hre from 'hardhat'
import { isFile, promptToContinue } from '@layerzerolabs/io-utils'
import { relative, resolve } from 'path'
import { TASK_LZ_WIRE_OAPP } from '@layerzerolabs/ua-utils-evm-hardhat'
import { deployOAppFixture } from '../../__utils__/oapp'
import { cwd } from 'process'

jest.mock('@layerzerolabs/io-utils', () => {
    const original = jest.requireActual('@layerzerolabs/io-utils')

    return {
        ...original,
        promptToContinue: jest.fn().mockRejectedValue('Not mocked'),
    }
})

const promptToContinueMock = promptToContinue as jest.Mock

describe('task/oapp/wire', () => {
    const CONFIGS_BASE_DIR = resolve(__dirname, '__data__', 'configs')
    const configPathFixture = (fileName: string): string => {
        const path = resolve(CONFIGS_BASE_DIR, fileName)

        expect(isFile(path)).toBeTruthy()

        return path
    }

    beforeEach(async () => {
        promptToContinueMock.mockReset()
    })

    describe('with invalid configs', () => {
        beforeAll(async () => {
            await deployOAppFixture()
        })

        it('should fail if the config file does not exist', async () => {
            await expect(hre.run(TASK_LZ_WIRE_OAPP, { oappConfig: './does-not-exist.js' })).rejects.toMatchSnapshot()
        })

        it('should fail if the config file is not a file', async () => {
            await expect(hre.run(TASK_LZ_WIRE_OAPP, { oappConfig: __dirname })).rejects.toMatchSnapshot()
        })

        it('should fail if the config file is not a valid JSON or JS file', async () => {
            const readme = resolve(__dirname, '..', '..', '..', 'README.md')

            expect(isFile(readme)).toBeTruthy()

            await expect(hre.run(TASK_LZ_WIRE_OAPP, { oappConfig: readme })).rejects.toMatchSnapshot()
        })

        it('should fail with an empty JSON file', async () => {
            const oappConfig = configPathFixture('invalid.config.empty.json')

            await expect(hre.run(TASK_LZ_WIRE_OAPP, { oappConfig })).rejects.toMatchSnapshot()
        })

        it('should fail with an empty JS file', async () => {
            const oappConfig = configPathFixture('invalid.config.empty.js')

            await expect(hre.run(TASK_LZ_WIRE_OAPP, { oappConfig })).rejects.toMatchSnapshot()
        })

        it('should fail with a malformed JS file (001)', async () => {
            const oappConfig = configPathFixture('invalid.config.001.js')

            await expect(hre.run(TASK_LZ_WIRE_OAPP, { oappConfig })).rejects.toMatchSnapshot()
        })

        it('should fail with a misconfigured file (001)', async () => {
            const oappConfig = configPathFixture('valid.config.misconfigured.001.js')

            await expect(hre.run(TASK_LZ_WIRE_OAPP, { oappConfig })).rejects.toMatchSnapshot()
        })
    })

    describe('with valid configs', () => {
        beforeEach(async () => {
            await deployOAppFixture()
        })

        it('should exit if there is nothing to wire', async () => {
            const oappConfig = configPathFixture('valid.config.empty.js')

            await hre.run(TASK_LZ_WIRE_OAPP, { oappConfig })

            expect(promptToContinueMock).not.toHaveBeenCalled()
        })

        it('should work with relative paths', async () => {
            const oappConfigAbsolute = configPathFixture('valid.config.empty.js')
            const oappConfig = relative(cwd(), oappConfigAbsolute)

            await hre.run(TASK_LZ_WIRE_OAPP, { oappConfig })

            expect(promptToContinueMock).not.toHaveBeenCalled()
        })

        it('should have debug output if requested (so called eye test, check the test output)', async () => {
            const oappConfig = configPathFixture('valid.config.connected.js')

            promptToContinueMock.mockResolvedValue(false)

            await hre.run(TASK_LZ_WIRE_OAPP, { oappConfig, logLevel: 'debug' })

            expect(promptToContinueMock).toHaveBeenCalledTimes(2)
        })

        it('should ask the user whether they want to see the transactions & continue', async () => {
            const oappConfig = configPathFixture('valid.config.connected.js')

            promptToContinueMock.mockResolvedValue(true)

            await hre.run(TASK_LZ_WIRE_OAPP, { oappConfig })

            expect(promptToContinueMock).toHaveBeenCalledTimes(2)
        })

        it('should return undefined if the user decides not to continue', async () => {
            const oappConfig = configPathFixture('valid.config.connected.js')

            promptToContinueMock.mockResolvedValue(false)

            const result = await hre.run(TASK_LZ_WIRE_OAPP, { oappConfig })

            expect(result).toBeUndefined()
            expect(promptToContinueMock).toHaveBeenCalledTimes(2)
        })

        it('should return a list of transactions if the user decides to continue', async () => {
            const oappConfig = configPathFixture('valid.config.connected.js')

            promptToContinueMock
                .mockResolvedValueOnce(false) // We don't want to see the list
                .mockResolvedValueOnce(true) // We want to continue

            const result = await hre.run(TASK_LZ_WIRE_OAPP, { oappConfig })

            expect(result).toEqual([])
        })
    })
})
