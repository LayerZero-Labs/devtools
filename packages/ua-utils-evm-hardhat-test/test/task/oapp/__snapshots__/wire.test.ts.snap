// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`task/oapp/wire with invalid configs should fail with a malformed JS file (001) 1`] = `
[Error: Config from file '/app/packages/ua-utils-evm-hardhat-test/test/task/oapp/__data__/configs/invalid.config.001.js' is malformed. Please fix the following errors:


contracts:
- Property 'contracts.0.contract': Invalid input
- Property 'contracts.1.contract': Invalid input,
connections:
- Property 'connections': Required]
`;

exports[`task/oapp/wire with invalid configs should fail with an empty JS file 1`] = `
[Error: Config from file '/app/packages/ua-utils-evm-hardhat-test/test/task/oapp/__data__/configs/invalid.config.empty.js' is malformed. Please fix the following errors:


contracts:
- Property 'contracts': Required,
connections:
- Property 'connections': Required]
`;

exports[`task/oapp/wire with invalid configs should fail with an empty JSON file 1`] = `[Error: Unable to read config file '/app/packages/ua-utils-evm-hardhat-test/test/task/oapp/__data__/configs/invalid.config.empty.json': SyntaxError: Unexpected end of JSON input]`;
